extensions.c

```diff
@@ -135,6 +225,104 @@ hibaext_decode(struct hibaext *ext, struct sshbuf *blob) {
 	return ret;
 }
 
+static int
+hibaext_decode_one(struct hibaext ***exts, int nexts, const unsigned char *data, size_t len) {
+	int ret;
+	struct sshbuf *blob = sshbuf_new();
+	sshbuf_put(blob, data, len);
+	*exts = xreallocarray(*exts, nexts + 1, sizeof(struct hibaext*));
+	(*exts)[nexts] = hibaext_new();
+	ret = hibaext_decode((*exts)[nexts], blob);
+	sshbuf_free(blob);
+	return ret;
+}
+
+int
+hibaext_decode_all(struct hibaext ***exts, struct sshbuf *blob) {
+	int nexts = 0;
+	int ret = 0;
+	size_t i = 0;
+	u_int32_t magic;
+	struct sshbuf *extensions = hibaext_maybe_inflate(blob);
+	const unsigned char *data = sshbuf_ptr(extensions);
+	size_t len = sshbuf_len(extensions);
+
+	if (data == NULL) {
+		ret = HIBA_INVALID_EXT;
+		goto err;
+	}
+
+	magic = PEEK_U32(data);
+	debug3("hibaext_decode_all: total extension len %zu, magic=0x%08x", len, magic);
+
+	if (magic == HIBA_B64_MAGIC) {
+		/* Base64 format.
+		 * Either one or multiple comma separated extensions. */
+		debug3("hibaext_decode_all: found base64 extension(s)");
+		int pos = 0;
+		while (i < len) {
+			int extension_len = 0;
+
+			/* If we find a comma or end of buffer, we should have
+			 * a full extension to parse. */
+			if (data[i] == ',') {
+				extension_len = i - pos;
+			} else if (i == len - 1) {
+				extension_len = len - pos;
+			}
+
+			if (extension_len > 0) {
+				debug3("hibaext_decode_all: found extension [%d, %zu]: %.*s", pos, i, extension_len, data+pos);
+				if ((ret = hibaext_decode_one(exts, nexts, data+pos, extension_len)) < 0) {
+					debug3("hibaext_decode_all: hibaext_decode_one returned %d: %s", ret, ssh_err(ret));
```

Comment:
Are decoding errors intentionally not treated as hard errors when calling hibaext_decode_all? The ret = nexts at the end overrides any error set here.

Reply:
Good catch. I assumed the code would use the goto err, but it breaks instead in order to run the proper free calls.

Done